plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/TestPinotIntegrationSmokeTest.java

```diff
@@ -578,12 +603,52 @@ public void testBrokerQueryWithTooManyRowsForSegmentQuery()
                 tooManyRowsTableValues);
     }
 
+    @Test
+    public void testMaxLimitForPassthroughQueries()
+            throws InterruptedException
+    {
+        assertQueryFails("SELECT string_col, updated_at_seconds" +
+                        "  FROM  \"SELECT updated_at_seconds, string_col FROM " + TOO_MANY_BROKER_ROWS_TABLE +
+                        "  LIMIT " + (MAX_ROWS_PER_SPLIT_FOR_BROKER_QUERIES + 1) + "\"",
+                "Broker query returned '13' rows, maximum allowed is '12' rows. with query \"select updated_at_seconds, string_col from too_many_broker_rows limit 13\"");
+
+        // Pinot issue preventing Integer.MAX_VALUE from being a limit: https://github.com/apache/incubator-pinot/issues/7110
+        assertQueryFails("SELECT * FROM \"SELECT string_col, long_col FROM " + ALL_TYPES_TABLE + " LIMIT " + Integer.MAX_VALUE + "\"",
+                "Unexpected response status: 500 for request \\{\"sql\" : \"select string_col, long_col from alltypes limit 2147483647\" \\} to url http://localhost:\\d+/query/sql, with headers \\{Accept=\\[application/json\\], Content-Type=\\[application/json\\]\\}, full response null");
+
+        // Pinot broker requests do not handle limits greater than Integer.MAX_VALUE
+        // Note that -2147483648 is due to an integer overflow in Pinot: https://github.com/apache/pinot/issues/7242
+        assertQueryFails("SELECT * FROM \"SELECT string_col, long_col FROM " + ALL_TYPES_TABLE + " LIMIT " + ((long) Integer.MAX_VALUE + 1) + "\"",
+                "Query select string_col, long_col from alltypes limit -2147483648 encountered exception org.apache.pinot.common.response.broker.QueryProcessingException@\\w+ with query \"select string_col, long_col from alltypes limit -2147483648\"");
+
+        String tooManyBrokerRowsTableValues = "VALUES ('string_0', '1620604800')," +
+                "  ('string_1', '1620604801')," +
+                "  ('string_2', '1620604802')," +
+                "  ('string_3', '1620604803')," +
+                "  ('string_4', '1620604804')," +
+                "  ('string_5', '1620604805')," +
+                "  ('string_6', '1620604806')," +
+                "  ('string_7', '1620604807')," +
+                "  ('string_8', '1620604808')," +
+                "  ('string_9', '1620604809')," +
+                "  ('string_10', '1620604810')," +
+                "  ('string_11', '1620604811')";
+
+        // Explicit limit is necessary otherwise pinot returns 10 rows.
```

Comment:
This behaviour might be very surprising for people looking at this from the Trino side. Not sure if we can do anything about this though?

Reply:
Yep, I think it was done in pinot to make data exploration easier, but it was not obvious until I actually ran tests and found that out.

Reply:
The broker limit helps with this: if no limit is supplied then the default broker limit is applied.